core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java

```diff
@@ -203,31 +199,39 @@ public void accept(Response<AiMessage> message) {
 
             AiMessage aiMessage = response.content();
 
-            if (context.hasChatMemory()) {
-                context.chatMemory(memoryId).add(response.content());
-            }
-
             if (!aiMessage.hasToolExecutionRequests()) {
+                if (context.hasChatMemory()) {
+                    context.chatMemory(memoryId).add(aiMessage);
+                }
                 break;
             }
```

Comment:
Why was this changed?

Reply:
because in case of hasToolExecutionRequest we need to replace the text with variable results in the AiMessage before adding it to the chatMemory

Reply:
Let's add a comment about that as it is definitely not clear